wire-grpc-client/src/main/java/com/squareup/wire/GrpcMethod.kt

```diff
@@ -211,6 +220,154 @@ internal sealed class GrpcMethod<S : Any, R : Any>(
     }
   }
 
+  class GrpcCall<S : Any, R : Any>(
+    path: String,
+    requestAdapter: ProtoAdapter<S>,
+    responseAdapter: ProtoAdapter<R>
+  ) : GrpcMethod<S, R>(path, requestAdapter, responseAdapter) {
+    override fun invoke(
+      grpcClient: GrpcClient,
+      args: Array<Any>
+    ): com.squareup.wire.GrpcCall<S, R> {
+      return object : com.squareup.wire.GrpcCall<S, R> {
+        // TODO(benoit) We cannot eagerly create it because we're waiting on the request.
+        private var call: Call? = null
+
+        // TODO(benoit) How do we deal with `Call` being lazily instantiated?
+        override val timeout: Timeout
+          get() = call!!.timeout()
+
+        override fun cancel() {
+          call?.cancel()
+        }
+
+        override suspend fun execute(request: S): R {
+          val requestBody = newRequestBody(requestAdapter, request)
+          call = grpcClient.newCall(path, requestBody)
+
+          return suspendCancellableCoroutine { continuation ->
+            continuation.invokeOnCancellation {
+              cancel()
+            }
+            call!!.enqueue(object : Callback {
+              override fun onFailure(call: Call, e: IOException) {
+                continuation.resumeWithException(e)
+              }
+
+              override fun onResponse(call: Call, response: Response) {
+                try {
+                  response.use {
+                    response.messageSource(responseAdapter).use { reader ->
+                      while (true) {
+                        val message = reader.read() ?: break
+                        continuation.resume(message)
+                      }
+
+                      val exception = response.grpcStatusToException()
+                      if (exception != null) continuation.resumeWithException(exception)
+                    }
+                  }
+                } catch (e: IOException) {
+                  continuation.resumeWithException(e)
+                }
+              }
+            })
+          }
+        }
+
+        override fun executeBlocking(request: S): R {
+          val requestBody = newRequestBody(requestAdapter, request)
+          call = grpcClient.newCall(path, requestBody)
+
+          val messageSource = BlockingMessageSource(call!!, responseAdapter)
+          call!!.enqueue(messageSource.readFromResponseBodyCallback())
+
+          messageSource.use {
+            return messageSource.read() ?: throw ProtocolException("required message not sent")
+          }
+        }
+
+        override fun enqueue(request: S, callback: com.squareup.wire.GrpcCall.Callback<S, R>) {
+          val requestBody = newRequestBody(requestAdapter, request)
+          call = grpcClient.newCall(path, requestBody)
+
+          val messageSource = BlockingMessageSource(call!!, responseAdapter)
+          call!!.enqueue(messageSource.readFromResponseBodyCallback())
+
+          messageSource.use {
+            try {
+              val response =
```

Comment:
I think this case wants us to check `response.grpcStatusToException()`.

Reply:
(or does BlockingMessageSource do that for us?)

Reply:
`BlockingMessageSource#readFromResponseBodyCallback` does it for us yes